src/lib/hooks/swap/useBestTrade.ts

```diff
@@ -0,0 +1,63 @@
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
+import { useClientSideV3Trade } from 'hooks/useClientSideV3Trade'
+import useDebounce from 'hooks/useDebounce'
+import { useMemo } from 'react'
+import { InterfaceTrade, TradeState } from 'state/routing/types'
+
+import useClientSideSmartOrderRouterTrade from '../routing/useClientSideSmartOrderRouterTrade'
+
+/**
+ * Returns the best v2+v3 trade for a desired swap.
+ * @param tradeType whether the swap is an exact in/out
+ * @param amountSpecified the exact amount to swap in/out
+ * @param otherCurrency the desired output/payment currency
+ */
+export function useBestTrade(
+  tradeType: TradeType,
+  amountSpecified?: CurrencyAmount<Currency>,
+  otherCurrency?: Currency
+): {
+  state: TradeState
+  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined
+} {
+  // debounce used to prevent excessive requests to SOR, as it is data intensive
+  // this helps provide a "syncing" state the UI can reference for loading animations
+  const [debouncedAmount, debouncedOtherCurrency] = useDebounce(
+    useMemo(() => [amountSpecified, otherCurrency], [amountSpecified, otherCurrency]),
+    200
+  )
+
+  const routingAPITrade = useClientSideSmartOrderRouterTrade(tradeType, debouncedAmount, debouncedOtherCurrency)
+
+  const isLoading = amountSpecified !== undefined && debouncedAmount === undefined
+
+  // consider trade debouncing when inputs/outputs do not match
+  const debouncing =
+    routingAPITrade.trade &&
+    amountSpecified &&
+    (tradeType === TradeType.EXACT_INPUT
+      ? !routingAPITrade.trade.inputAmount.equalTo(amountSpecified) ||
+        !amountSpecified.currency.equals(routingAPITrade.trade.inputAmount.currency) ||
+        !debouncedOtherCurrency?.equals(routingAPITrade.trade.outputAmount.currency)
+      : !routingAPITrade.trade.outputAmount.equalTo(amountSpecified) ||
+        !amountSpecified.currency.equals(routingAPITrade.trade.outputAmount.currency) ||
+        !debouncedOtherCurrency?.equals(routingAPITrade.trade.inputAmount.currency))
```

Comment:
can you simplify this?

Reply:
yeah this is leftover from app and agreed its brutal to read. Will attempt to clean up

Reply:
yeah, figured that was prob the case so just gave it an approval anyway

if you feel like fixing I will love you forever ❤️

Reply:
this seems like it is examining effects to determine if debounce has occurred. Can you rewrite the debounce to keep state so you don't need this complex logic?